core/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -792,6 +793,17 @@ private void parseSource(DefaultSearchContext context, SearchSourceBuilder sourc
             }
             context.storedFieldsContext(source.storedFields());
         }
+
+        if (source.collapse() != null) {
```

Comment:
I've changed it so now we add the field directly in DocValueFetchSubPhase if the collapse field is defined.
It's not what you proposed but I think it's cleaner, WDYT